cli/cli/kurtosis_gateway/connection/connection.go

```diff
@@ -89,17 +87,56 @@ func newLocalPortToPodPortConnection(kubernetesRestConfig *k8s_rest.Config, podP
 		Timeout:       0,
 	}, dialerMethod, podProxyEndpointUrl)
 
-	// Create our k8s port forwarder
-	portForwarder, err := portforward.NewOnAddresses(dialer, portForwardAddresses, portStrings, portforwardStopChannel, portforwardReadyChannel, &portforwardStdOut, &portforwardStdErr)
-	if err != nil {
-		return nil, err
-	}
-	// Start forwarding ports asynchronously
+	// Start forwarding ports asynchronously with reconnect logic.
+	// The reconnect logic tries to reconnect after a working connection is lost.
+	// The port forward process can be interrupted using the port forwarder stop channel.  There is no retry after that.
+	var portForwarder *portforward.PortForwarder
 	go func() {
-		if err := portForwarder.ForwardPorts(); err != nil {
-			logrus.Warnf("Expected to be able to start forwarding local ports to remote ports, instead our portforwarder has returned a non-nil err:\n%v", err)
+		retries := 0
+		readyChannel := portforwardReadyChannel
+		for {
+			portForwarder, err = portforward.NewOnAddresses(dialer, portForwardAddresses, portStrings, portforwardStopChannel, readyChannel, &portforwardStdOut, &portforwardStdErr)
+			if err != nil {
+				// Addresses or ports cannot be parsed so there is nothing else to try
+				logrus.Errorf("An error occured parsing the port forwarder addresses or ports:\n%v", err)
+				return
+			} else {
+				logrus.Debugf("Opening connection to pod: %s", podProxyEndpointUrl.String())
+				if err = portForwarder.ForwardPorts(); err != nil {
+					if err == portforward.ErrLostConnectionToPod {
+						logrus.Infof("Lost connection to pod: %s", podProxyEndpointUrl.String())
+						// Copy the port forwarder assigned local ports so we re-use the same local ports when we reconnect
+						ports, err := portForwarder.GetPorts()
+						if err != nil {
+							logrus.Errorf("An error occured retrieving the local ports to remote ports mapping for our portforwarder:\n%v", err)
+							return
+						}
+						portStrings = nil
+						for _, port := range ports {
+							portString := fmt.Sprintf("%v:%v", port.Local, port.Remote)
+							portStrings = append(portStrings, portString)
+						}
+					} else {
+						if retries == 0 {
+							// Exit the retry logic if the first try to connect fails
+							logrus.Errorf("Expected to be able to start forwarding local ports to remote ports, instead our portforwarder has returned a non-nil err:\n%v", err)
+							return
+						}
+						logrus.Debugf("Error trying to forward ports:\n%v", err)
+					}
```

Comment:
@laurentluce I'm probably missing something here but I struggle with this logic:
If the error is not `ErrLostConnectionToPod` and retries >0 we then basically continue to retry indefinitely? Is that the intention? 

What types of errors would we get if not `ErrLostConnectionToPod`? My assumption is that it'll be a set of more or less unrecoverable issues.  I guess I would expect the logic to be: If error is not `ErrLostConnectionToPod` we log an error.

Let me know your thoughts

Reply:
@adschwartz `ErrLostConnectionToPod` is returned when a previously working connection breaks.  Other errors will be connection initiation failures.  Once the connection breaks, the logic tries to reconnect every 5 seconds and we are going to get connection initiation failures until the pod is accessible again.  I chose the debug level here to not flood the gateway output until we are able to reconnect.